README
# Mythbuntu Control Centre
# By: Mario Limonciello

---                                  ---
-- What's this wonderful application? --
---                                  ---
This application is used for managing administration tasks on a Mythbuntu
system.  It aims to remove the command line requirement typically involved
with working on a Mythbuntu box.

It has been developed with a "Fully Pluggable" architecture.  This means
that all functionality is actually configurable by the plugins that get
installed in the standard location rather than with the core application.

This allows developers to create their own sets of plugins to be used
with the Control Centre.

---                    ---
-- Architecture Defined --
---                    ---
The MCC (aka Mythbuntu Control Centre) is a client/server application
with the server and client intended to run on the same machine.  The two
components contact one another via dbus with authentication from
PolicyKit.  This allows the frontend and all GUI functionality to run
as a userspace process.  The backend is spawned using dbus service
activation as necessary.  When inactive for a pre-defined time, the
backend will stop itself and only respawn when requested by the
frontend.

The intention behind this architecture definition is to abstract the
developer from having to spend time re-inventing processes that are
likely already in use by other MCC plugins.

The core MCC application provides hooks available for all plugins to
use. Optionally these hooks can define post install actions or apt
related actions.

Functionally, the frontend and backend processes will both import
the python plugin and initialize the class for each process.  The
frontend also calls the GTK functions, whereas the backend only calls
upon two backend functions.

All plugins are stored in /usr/share/mythbuntu/plugins/{python,ui}.
Each plugin requires a python source file and a GtkBuilder UI file.
The details for how these need to be implemented are defined below.

---                ---
-- Writing a plugin --
---                ---
An example plugin is included with the distribution of the core app.
While explaining how to write a plugin, we will be referring to this
plugin.

Familiarize yourself with the plugin by running this command:
 # mythbuntu-control-centre --plugin-root-path=/usr/share/mythbuntu/examples/plugins

This launches the control centre with only plugins found in that
plugin-root-path.  This is helpful for development so that you know that
your plugin is not causing problems early on.

The UI will display and you can see that this basic plugin will show the
following elements:
 - Installing & Removing a package
 - Processing a scripted activity as root
 - Processing a scripted activity as a user
 - Using a GTK callback to change elements of the UI

After you are familiar with how this plugin works, you can copy it into
a home directory to start working on your own plugin with it as a basis.
Make sure that you call it with the full path as an argument.

It's best to develop the frontend of a plugin before the backend.  Start
out using the Glade Development tool, glade-3.

Opening up glade-3, a few items should be apparent.
 - The plugin's file name needs to match the top most non window widget
   of the plugin.  This is what is used to key off what file the plugin
   loads for the GUI.
 - The goal is to use the minimum amount of space.  Try not to add too
   much text as the control centre's minimum size is determined by the
   max size of a plugin.
 - Use alignment boxes to keep widgets from the edges.  They look much
   better this way.

There are tons of advanced features you can add to your ui file, but
those are out of the scope of this guide.  Look around the internet for
such guides.

After finishing off the GUI, be sure to take note of any widgets that
you will need to be keying off in the python file.

Open up the python file in your favorite editor.  We'll discuss the
elements that are required for the frontend of the plugin first.

--Frontend--

A frontend plugin will always inherit from the class MCCPlugin.  By doing
so, a variety of methods will be available for the plugin.  You'll need
to override these methods at a minimum to ensure proper functionality:
 - __init__
 - captureState
 - applyStateToGUI
 - compareState

In __init__ you need to define a dictionary with the items 'name', 'icon'
and 'ui'.  After building this dictionary, you need to call the parent
MCCPlugin __init__ with that dictionary.
 - 'name' is the name of the plugin on the right side of the MCC UI
 - 'icon' is the icon that will show up on the right side of the UI
 - 'ui' is the name of the GtkBuilder file to be loaded (sans .ui)

captureState captures the state of all elements on the system.  It is
intentionally unpaired with applyStateToGUI and compareState because
MCC may call these at any time.  It's best to store any information
determined about the installed system in a dictionary for later use.
 - query_installed can be used for querying packaged applications
 - you can import any python packages and use them as well

applyStateToGUI will override any currently set GUI elements with things
that were determined in captureState.

compareState will compare that dictionary with the currently set GUI
elements to determine what's changed.
 - If it is determined that the page needs more activity before
   being "done", self._incomplete can be set to True.
 - It's important to call MCCPlugin.clearParentState(self) in this
   function so that the frontend is in a standard state.

The following functions can be used to record what the backend needs:
 - _markInstall: marks a package to be installed
 - _markRemove: marks a package to be removed
 - _markReconfigureUser: marks a scripted change to be done as a user
 - _markReconfigureRoot: marks a scripted change to be done by root
 - _markUpdatePackageList : requests a package list update after completion
 - _markUnauthenticatedPackages : requests that an unauthenticated package be installed

Callbacks can also be used, but generally the model is that changes
shouldn't occur until after the frontend calls apply.

--Backend--

If you are only making package installs or removals, you don't need to
define any more functionality to the plugin.

If you need to represent changes that aren't part of the package
(scriptedchanges), then you need to define two more methods:
 - root_scripted_changes
 - user_scripted_changes
Both methods have an argument of a dictionary of items to change

The exact same python source file is loaded into the backend process
when spawned if necessary, so be cognizant that you shouldn't do any
GTK initialization in __init__ or the backend will fail.

Generally you want to walk through the dictionary argument for both
cases through a for loop as more than one item can be sent at a time.

Running 'dpkg-reconfigure PACKAGE' is OK in this section, but be sure
to set the env to 'noninteractive' and instead make the changes in flat
text files that can be read by the maintainer scripts of that package.

Try to use subprocess.call rather than os.system so the return status
can be determined.

If you are running an application that may take a long time in your
processing functions, you might want to use emit_progress which will
update the GUI with the latest status.

To assist in debugging, you may consider importing the 'logging' module.
logging.warning('message') will always show up in the backend log
logging.debug('message') will show up when the backend is spawned with
--debug.

--Packaging--
When your plugin is stable, you can start moving it into the system
location of /usr/share/mythbuntu/plugins.  This means that the standard
invokation of mythbuntu-control-centre will initialize your plugin
along with the others on the system.  This adds a new set of variables,
particularly if you have any name clashes.  Be sure to test like this
before issuing your plugin to make sure there are no blatant problems.


--Single Plugin Mode--
Sometimes plugins can be so complex that they make more sense being
put out on their own, but fit very well into the features that MCC provides.

In these instances, you can run the application in "single plugin mode".
If the plugin should still be shown in the UI in all instances, then
install it to the standard location and run the plugin like this:

#mythbuntu-control-centre -S "Log Grabber"

where "Log Grabber" is the name of one of the plugins.

If the plugin should only be shown in single plugin mode, install it elsewhere
on the system, and just set the plugin-root-path:

#mythbuntu-control-centre --plugin-root-path=/usr/share/mythbuntu/examples/plugins/ -s "Skeletor"

You can place whatever you decide into a .desktop file to install into the menus.


debian/changelog
mythbuntu-control-centre (0.64.1) trusty; urgency=medium

  * Rewrite the python shebang path to /usr/bin/python3.

 -- Matthias Klose <doko@ubuntu.com>  Thu, 20 Feb 2014 16:18:25 +0100

mythbuntu-control-centre (0.64) trusty; urgency=low

  [ Thomas Mashos ]
  * Sort plugins alphabetically
  * Fix running with Python3
  * One item per line in summarizing

  [ Mario Limonciello ]
  * debian/control:
    - Update for python3 stuff.
    - drop cdbs
    - update for current lintian.
  * debian/rules:
    - drop cdbs stuff
    - don't make separate upstream and ubuntu versions.
  * Correct an enum for APT installation.

 -- Mario Limonciello <superm1@ubuntu.com>  Tue, 12 Nov 2013 01:15:52 -0600

mythbuntu-control-centre (0.63-0ubuntu2) oneiric; urgency=low

  * langpack.mk got dropped from cdbs in favour of dh_translations - LP: #833884

 -- Marcin Juszkiewicz <marcin.juszkiewicz@linaro.org>  Wed, 31 Aug 2011 16:54:42 +0000

mythbuntu-control-centre (0.63-0ubuntu1) natty; urgency=low

  * Update commit_packages routine for changes in aptdaemon.
  * Set the apt daemon dialog to always center.

 -- Mario Limonciello <Mario_Limonciello@Dell.com>  Mon, 07 Feb 2011 23:54:25 -0600

mythbuntu-control-centre (0.62-0ubuntu2) natty; urgency=low

  * debian/control:
    - Update bzr branches to point to ~mythbuntu-dev.
  * Rebuild for python2.7 transition.

 -- Mario Limonciello <Mario_Limonciello@Dell.com>  Wed, 02 Feb 2011 18:15:00 -0600

mythbuntu-control-centre (0.62-0ubuntu1) lucid; urgency=low

  * Correct some deprecation warnings from a recent python-apt upload.

 -- Mario Limonciello <Mario_Limonciello@Dell.com>  Sun, 28 Mar 2010 01:21:16 -0500

mythbuntu-control-centre (0.61-0ubuntu1) lucid; urgency=low

  * Widget names are no longer stored in the widget's internal name, but
    rather in the object data.  Sync this information back to the widget's
    internal name. (LP: #503710)

 -- Mario Limonciello <superm1@ubuntu.com>  Wed, 06 Jan 2010 03:23:08 -0600

mythbuntu-control-centre (0.60-0ubuntu1) lucid; urgency=low

  * Update for the API changes with python-aptdaemon.
  * Update for changes with PolicyKit.

 -- Mario Limonciello <superm1@ubuntu.com>  Tue, 05 Jan 2010 01:00:02 -0600

mythbuntu-control-centre (0.59-0ubuntu1) lucid; urgency=low

  * Introduce a "single" plugin mode.  This allows a plugin to be launched
    without allowing switching to other plugins in the UI.

 -- Mario Limonciello <superm1@ubuntu.com>  Fri, 27 Nov 2009 17:28:59 -0600

mythbuntu-control-centre (0.58-0ubuntu1) lucid; urgency=low

  * Add a new function to all plugins for notification during extended
    processing.
  * Update skeletor to show how to update the progress bar from a plugin.
  * Update policykit to cache between runs.

 -- Mario Limonciello <superm1@ubuntu.com>  Wed, 25 Nov 2009 02:17:53 -0600

mythbuntu-control-centre (0.57-0ubuntu1) karmic; urgency=low

  * When plugins get loaded into the button box, pack them into gtk.HBox's
    rather than using gtk.Button.set_image.  It appears that gtk-button-images
    is not set on Gnome installs causing different behaviors on Gnome vs.
    XFCE otherwise.
  * Drop diverstion of mythtv menu xml file to add our entry in.
  * Drop old postinst that is not needed for upgrade scenarios anymore.

 -- Mario Limonciello <superm1@ubuntu.com>  Wed, 14 Oct 2009 01:24:28 -0500

mythbuntu-control-centre (0.56-0ubuntu2) karmic; urgency=low

  * debian/control:
    - Fix small comma error that didn't throw off pbuilder, but sbuild didn't
      like.

 -- Mario Limonciello <superm1@ubuntu.com>  Sun, 04 Oct 2009 04:09:21 -0500

mythbuntu-control-centre (0.56-0ubuntu1) karmic; urgency=low

  * debian/dirs:
    - Install ui rather than glade directory.
  * ui:
    - Update default window size from 800x400 to 790x480.
  * Use aptdaemon for performing package install and removals
    rather than complex built-in logic.  (Stop reinventing the square!)
  * Allow plugins to request a package list update so that future
    invokations will not have errors.
  * Allow a plugin to request that a single package transaction allow
    unauthenticated installation.  A large WARNING is presented to the
    user when this option is enabled by a plugin. (LP: #415969)

 -- Mario Limonciello <superm1@ubuntu.com>  Sun, 04 Oct 2009 04:00:25 -0500

mythbuntu-control-centre (0.55-0ubuntu1) karmic; urgency=low

  * Port to polkit-1.
  * debian/control:
    - Depends on policykit-1 instead of policykit

 -- Mario Limonciello <superm1@ubuntu.com>  Sat, 03 Oct 2009 16:24:39 -0500

mythbuntu-control-centre (0.54-0ubuntu1) karmic; urgency=low

  * Set notebook tab focus on click. (LP: #428689)
  * Don't default one of the tabs to be focused when welcome message
    is shown. (LP: #428691)

 -- Mario Limonciello <superm1@ubuntu.com>  Mon, 14 Sep 2009 23:06:08 -0500

mythbuntu-control-centre (0.53-0ubuntu1) karmic; urgency=low

  * mythbuntu_control_centre.ui:
    - Don't let dialog boxes be named the same.  A recent GTK upload
      appears to be less forgiving of it. (LP: #422240)
  * debian/control:
    - Depends on python-dbus

 -- Mario Limonciello <superm1@ubuntu.com>  Mon, 31 Aug 2009 22:36:33 -0500

mythbuntu-control-centre (0.52-0ubuntu1) karmic; urgency=low

  * Convert the project to use gtk.Builder as the primary way of loading
    plugins.  This introduces a new information dictionary key, "ui"
    for plugins. (LP: #403542)
  * Add more debugging code when loading widgets and files and what not
    as they have more of a tendency to break now.
  * debian/control:
    - Remove python-glade2
    - Add python-gtk2

 -- Mario Limonciello <superm1@ubuntu.com>  Tue, 18 Aug 2009 00:09:51 -0500

mythbuntu-control-centre (0.51-0ubuntu1) karmic; urgency=low

  * applications/mythbuntu-control-centre.desktop.in:
    - Set the icon to the proper icon so we can see it in XFCE menus.

 -- Mario Limonciello <superm1@ubuntu.com>  Wed, 10 Jun 2009 12:07:20 -0500


examples/plugins/python/skeletor.py
## -*- coding: utf-8 -*-
#
# «skeletor» - An Example Plugin for showing how to use MCC as a developer
#
# Copyright (C) 2009, Mario Limonciello, for Mythbuntu
#
#
# Mythbuntu is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free
# Software Foundation; either version 2 of the License, or at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this application; if not, write to the Free Software Foundation, Inc., 51
# Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
##################################################################################

from MythbuntuControlCentre.plugin import MCCPlugin
from gi.repository import Gtk
import time

class SkeletonPlugin(MCCPlugin):
    """An Example Plugin for showing how to use MCC as a developer"""
    #
    #Load GUI & Calculate Changes
    #

    def __init__(self):
        #Initialize parent class
        information = {}
        information["name"] = "Skeletor"
        information["icon"] = "gtk-stop"
        information["ui"] = "tab_skeletor"
        MCCPlugin.__init__(self,information)

    def captureState(self):
        """Determines the state of the items on managed by this plugin
           and stores it into the plugin's own internal structures"""
        import os
        self.changes = {}
        self.changes['user_file'] = os.path.exists('/tmp/user_file')
        self.changes['root_file'] = os.path.exists('/tmp/root_file')
        self.changes['tuxeyes'] = self.query_installed('tuxeyes')

    def applyStateToGUI(self):
        """Takes the current state information and sets the GUI
           for this plugin"""
        self.user_touch_checkbox.set_active(self.changes['user_file'])
        self.root_touch_checkbox.set_active(self.changes['root_file'])
        self.tuxeyes_checkbox.set_active(self.changes['tuxeyes'])

    def compareState(self):
        """Determines what items have been modified on this plugin"""
        MCCPlugin.clearParentState(self)
        if self.user_touch_checkbox.get_active() != self.changes['user_file']:
            self._markReconfigureUser('user_file',self.user_touch_checkbox.get_active())
        if self.root_touch_checkbox.get_active() != self.changes['root_file']:
            self._markReconfigureRoot('root_file',self.root_touch_checkbox.get_active())
        if self.tuxeyes_checkbox.get_active() != self.changes['tuxeyes']:
            if self.tuxeyes_checkbox.get_active():
                self._markInstall('tuxeyes')
            else:
                self._markRemove('tuxeyes')
    #
    # Callbacks
    #
    def callback_example(self,widget,data=None):
        """Shows an example of how a callback can do stuff"""
        widget_was_visible = self.callback_image.flags() & Gtk.VISIBLE
        if widget_was_visible:
            self.callback_image.hide()
        else:
            self.callback_image.show()

    #
    # Process selected activities
    #

    def root_scripted_changes(self,reconfigure):
        """System-wide changes that need root access to be applied.
           This function is ran by the dbus backend"""
        self.emit_progress("Setting to 20 percent as a root", 20)
        time.sleep(2)
        for item in reconfigure:
            if item == "root_file":
                if reconfigure[item]:
                    file = open("/tmp/root_file", "a")
                else:
                    import os
                    os.remove("/tmp/root_file")
        self.emit_progress("Setting to 90 percent as a root", 90)
        time.sleep(2)

    def user_scripted_changes(self,reconfigure):
        """Local changes that can be performed by the user account.
           This function will be ran by the frontend"""
        self.emit_progress("Setting to 20 percent as a user", 20)
        time.sleep(2)
        for item in reconfigure:
            if item == "user_file":
                if reconfigure[item]:
                    file = open("/tmp/user_file", "a")
                else:
                    import os
                    os.remove("/tmp/user_file")
        self.emit_progress("Setting to 90 percent as a user", 90)
        time.sleep(2)


MythbuntuControlCentre/plugin.py
## -*- coding: utf-8 -*-
#
# «plugin» - The parent level class that all MCC plugins inherit from
#
# Copyright (C) 2009, Mario Limonciello, for Mythbuntu
#
#
# Mythbuntu is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free
# Software Foundation; either version 2 of the License, or at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this application; if not, write to the Free Software Foundation, Inc., 51
# Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
##################################################################################

import logging
import sys
import os
import string
import traceback

#GUI Support
from gi.repository import Gtk

#Translation Support
from gettext import gettext as _

class MCCPluginLoader():
    """A class used for initializing all loadable plugins"""
    def __init__(self,plugin_root_path):
        self._instances = {}

        self.plugin_root_path = plugin_root_path
        self.plugin_path = plugin_root_path + '/python'

        if self.plugin_path not in sys.path:
            sys.path.insert(0, self.plugin_path)

    def reload_plugins(self):
        """Searches the path for all plugins asked for"""
        plugins=[]
        for obj in os.listdir(self.plugin_path):
            if '.py' in obj and '.pyc' not in obj:
                plugins.append(obj.split('.py')[0])

        for plugin in plugins:
            logging.debug(_("Importing plugin") + ": \t" + plugin)
            try:
                __import__(plugin, None, None, [''])
            except:
                logging.warning( _("Error importing plugin ") + plugin)
                traceback.print_exc()
                continue

    def find_plugin_classes(self):
        """Returns currently loaded plugin class types"""
        logging.debug(_("Found the following plugin classes:"))
        logging.debug(MCCPlugin.__subclasses__())

        return MCCPlugin.__subclasses__()

    def find_plugin_instances(self):
        """Returns all current instances of plugins"""
        result = []
        for plugin in self.find_plugin_classes():
            if not plugin in self._instances:
                try:
                    self._instances[plugin] = plugin()
                    self._instances[plugin]._set_root_path(self.plugin_root_path)
                except:
                    logging.warning( _("Error loading plugin ") + str(plugin))
                    traceback.print_exc()
                    continue
            result.append(self._instances[plugin])

        logging.debug(_("Found the following plugin instances:"))
        logging.debug(result)

        return result

class MCCPlugin(object):
    """An abstract class that defines what all plugins need to be able to do"""

    def __init__(self,information):
        """Initializes a parent MCC plugin.  Information is expected to be a dictionary containing
           these keys at a minimum:
            name: Name of the plugin
            icon: Icon shown for the plugin

            And one of these:
            ui: UI file used for the plugin"""
        if "name" not in information or \
           "icon" not in information or \
           "ui" not in information:
            self._abstract("__init__: information keys")
        self._information = information
        self._incomplete = False
        self.clearParentState()

    ###Helper functions###
    def _set_root_path(self,plugin_root_path):
        """Sets the path to load ui from"""
        self.plugin_root_path = plugin_root_path

    def _mark_array(self,array,item,value,action):
        """Internal helper function for modifying arrays"""
        if action:
            if type(array) is dict:
                array[item]=value
            else:
                array.append(item)
        else:
            if type(array) is dict:
                array.pop(item)
            else:
                for tmp in array:
                    if tmp == item:
                        array.remove(item)
                        return

    def _abstract(self, method):
        """Common error that gets raised if a plugin does not redefine
           a method that is supposed to be"""
        raise NotImplementedError("%s.%s does not implement %s" %
                                  (self.__class__.__module__,
                                   self.__class__.__name__, method))

    def updateCache(self,cache):
        """Updates the apt package cache"""
        self.pkg_cache=cache

    def getInformation(self,key=None):
        """Returns a standard information key"""
        if key is None:
            return self._information
        elif key is 'module':
            return self.__class__.__module__
        else:
            return self._information[key]

    def query_installed(self,package):
        """Determines if a single package is installed"""
        try:
            result = self.pkg_cache[package].current_ver
            if result == None:
                return False
            else:
                return True
        except KeyError:
            return False

    def getIncomplete(self):
        """Returns whether a plugin has been fully filled out"""
        return self._incomplete

    def launch_app(self,widget,data=None):
        """Launches an external application"""
        if widget is not None and data is not None:
            parent=widget.get_parent_window().get_toplevel()
            parent.hide()
            while Gtk.events_pending():
                Gtk.main_iteration()
            os.system(data)
            parent.show()
        else:
            self._abstract("launch_app")

    def insert_extra_widgets(self):
        """Litters the namespace with extra widgets if they exist.
           Generally these would be around for things like popup
           windows that are plugin specific"""

        ui_file = os.path.join(self.plugin_root_path, 'ui', self._information["ui"] + ".extra.ui")
        if os.path.exists(ui_file):
            logging.debug("Reading UI file: %s" % ui_file)
            self.builder.add_from_file(ui_file)

    def insert_subpage(self,notebook,buttonbox,handler):
        """Inserts a subtab into the notebook.  This assumes the file
        shares the same base name as the page you are looking for.
        Returns tuple: (name,tab) where tab is the numeric index of the
        tab in the GtkNoteBook"""

        # Button for the notebook widget
        label=Gtk.Label(label=self._information["name"])
        icon=Gtk.Image()
        icon.set_from_icon_name(self._information["icon"],3)
        button=Gtk.Button()
        button.set_alignment(0,0)
        button.set_relief(Gtk.ReliefStyle.NONE)
        button.set_focus_on_click(True)
        hbox=Gtk.HBox(spacing=2)
        hbox.pack_start(icon,False,False,0)
        hbox.pack_start(label,False,False,0)
        button.add(hbox)
        buttonbox.add(button)
        button.connect("clicked",handler)
        label.show()
        icon.show()
        hbox.show()
        button.show()

        # See http://faq.pyGtk.org/index.py?req=show&file=faq22.002.htp
        # for internationalisation support
        widget = None
        self.builder = Gtk.Builder()
        ui_file = os.path.join(self.plugin_root_path,'ui',self._information["ui"] + ".ui")
        logging.debug("Reading .ui file: %s" % ui_file)
        self.builder.add_from_file(ui_file)
        self.builder.connect_signals(self)
        for widget in self.builder.get_objects():
            if not isinstance(widget, Gtk.Widget):
                continue
            widget.set_name(Gtk.Buildable.get_name(widget))
            setattr(self, widget.get_name(), widget)
        #widget that we will append in the notebook
        widget = self.builder.get_object(self._information["ui"])

        logging.debug("Appending Widget: %s" % widget.get_name())
        notebook.append_page(widget,None)
        return (self._information["name"],notebook.page_num(widget))

    ###State machine of the plugin###
    def clearParentState(self):
        """Clears the state of the elements that were stored for between
           runs"""
        self._to_install     = []
        self._to_remove      = []
        self._to_reconfigure_root = {}
        self._to_reconfigure_user = {}
        self._request_update = False
        self._request_unauth = False

    def captureState(self):
        """Determines the state of the items on managed by this plugin
           and stores it into the plugin's own internal structures"""
        self._abstract("captureState")

    def compareState(self):
        """Determines what items have been modified on this plugin"""
        self._abstract("compareState")

    def applyStateToGUI(self):
        """Takes the current state information and sets the GUI
           for this plugin"""
        self._abstract("applyStateToGUI")

    ###Interfacing functions###
    def _markInstall(self,item,install=True):
        """Modifies the mark of an item to be installed"""
        self._mark_array(self._to_install,item,None,install)

    def _markRemove(self,item,remove=True):
        """Modifies the mark of an item to be removed"""
        self._mark_array(self._to_remove,item,None,remove)

    def _markUnauthenticatedPackages(self,unauth=True):
        """Toggles the bit to request an unauthenticated package to be installed"""
        self._request_unauth = unauth

    def _markUpdatePackageList(self,update=True):
        """Toggles the bit that requests a package list update at the end"""
        self._request_update = update

    def _markReconfigureRoot(self,item,value,reconfigure=True):
        """Modifies the mark of an item to be removed"""
        self._mark_array(self._to_reconfigure_root,item,value,reconfigure)

    def _markReconfigureUser(self,item,value,reconfigure=True):
        """Modifies the mark of an item to be removed"""
        self._mark_array(self._to_reconfigure_user,item,value,reconfigure)

    def getRawChanges(self):
        """Returns a tupple of raw arrays that can be assembled together"""
        return (self._to_install,self._to_remove,self._to_reconfigure_root, self._to_reconfigure_user,self._request_update,self._request_unauth)

    def summarizeChanges(self):
        """Returns a pretty summary of all management activities that will occur"""
        def summarizeDictionary(array,action):
            """Returns a summary of what's happening in array"""
            text=''
            if len(array) > 0:
                text+=action + " the following items:\n"
                for item in array:
                    text += '\t' + item + '\n'
            return text
        summary=summarizeDictionary(self._to_install,"Install") + \
                summarizeDictionary(self._to_remove, "Remove")  + \
                summarizeDictionary(self._to_reconfigure_root, "Reconfigure (as root)") + \
                summarizeDictionary(self._to_reconfigure_user, "Reconfigure (as user)")
        if self._request_update:
            summary = summary + "Request Package List Update\n"
        if self._request_unauth:
            summary = summary + "*** WARNING ***\n Unauthenticated Packages Will be Installed during this transaction.\n *** WARNING ***\n"
        if summary == '':
            summary=False
        return summary

    ###Non package-able changes###
    def emit_progress(self, string, pct):
        """Emits a progress event through the backend over dbus"""

    def root_scripted_changes(self,reconfigure):
        """System-wide changes that need root access to be applied.
           This function is ran by the dbus backend"""
        self._abstract("root_scripted_changes")

    def user_scripted_changes(self,reconfigure):
        """Local changes that can be performed by the user account.
           This function will be ran by the frontend."""
        self._abstract("user_scripted_changes")
